File: examples/mini-ts/src/codegen.ts
import {
    BinaryExpression,
    CallExpression,
    Expression,
    ExpressionStatement,
    FunctionDeclaration,
    Identifier,
    IfStatement,
    Literal,
    literalToGaiaConstant,
    Program,
    ReturnStatement,
    Statement,
    typeToGaiaType,
    UnaryExpression,
    VariableDeclaration,
    WhileStatement
} from './ast';

// 假设我们有一个gaia-frontend-wasm32库可以导入
// 在实际实现中，这里会导入实际的Gaia类型和函数
// import { GaiaInstruction, GaiaProgram, GaiaFunction, GaiaType } from 'gaia-frontend-wasm32';

// 为了示例，我们先定义一些基本的Gaia类型
class GaiaInstruction {
    constructor(public opcode: string, public operands: any[] = []) {
    }
}

class GaiaFunction {
    constructor(
        public name: string,
        public parameters: { name: string; type: string }[],
        public returnType: string,
        public instructions: GaiaInstruction[]
    ) {
    }
}

class GaiaProgram {
    constructor(public functions: GaiaFunction[]) {
    }
}

export class CodeGenerator {
    private program: GaiaProgram;
    private currentFunction?: GaiaFunction;
    private localVariables: Map<string, number> = new Map();
    private localIndex: number = 0;

    constructor() {
        this.program = new GaiaProgram([]);
    }

    public generate(program: Program): GaiaProgram {
        this.program = new GaiaProgram([]);

        for (const func of program.functions) {
            this.generateFunction(func);
        }

        return this.program;
    }

    private generateFunction(func: FunctionDeclaration): void {
        // 重置局部变量状态
        this.localVariables = new Map();
        this.localIndex = 0;

        // 创建参数映射
        for (let i = 0; i < func.parameters.length; i++) {
            this.localVariables.set(func.parameters[i].name, i);
        }

        // 生成函数体
        const instructions: GaiaInstruction[] = [];

        for (const stmt of func.body) {
            instructions.push(...this.generateStatement(stmt));
        }

        // 如果函数没有返回语句且返回类型不是void，添加默认返回
        if (func.returnType !== 'void' &&
            (!func.body.length || func.body[func.body.length - 1].kind !== 'ReturnStatement')) {
            instructions.push(new GaiaInstruction('Push', [literalToGaiaConstant({
                kind: 'Literal',
                type: func.returnType,
                value: func.returnType === 'boolean' ? false : 0
            })]));
            instructions.push(new GaiaInstruction('Return'));
        }

        // 创建Gaia函数
        const gaiaFunc = new GaiaFunction(
            func.name,
            func.parameters.map(p => ({name: p.name, type: typeToGaiaType(p.type)})),
            typeToGaiaType(func.returnType),
            instructions
        );

        this.program.functions.push(gaiaFunc);
    }

    private generateStatement(stmt: Statement): GaiaInstruction[] {
        switch (stmt.kind) {
            case 'VariableDeclaration':
                return this.generateVariableDeclaration(stmt);
            case 'ReturnStatement':
                return this.generateReturnStatement(stmt);
            case 'IfStatement':
                return this.generateIfStatement(stmt);
            case 'WhileStatement':
                return this.generateWhileStatement(stmt);
            case 'ExpressionStatement':
                return this.generateExpressionStatement(stmt);
            default:
                throw new Error(`Unsupported statement kind: ${stmt.kind}`);
        }
    }

    private generateVariableDeclaration(stmt: VariableDeclaration): GaiaInstruction[] {
        const instructions: GaiaInstruction[] = [];

        if (stmt.initializer) {
            instructions.push(...this.generateExpression(stmt.initializer));
        }

        // 分配局部变量索引
        const index = this.localIndex++;
        this.localVariables.set(stmt.name, index);

        // 如果有初始值，存储到局部变量
        if (stmt.initializer) {
            instructions.push(new GaiaInstruction('StoreLocal', [index]));
        }

        return instructions;
    }

    private generateReturnStatement(stmt: ReturnStatement): GaiaInstruction[] {
        const instructions: GaiaInstruction[] = [];

        if (stmt.value) {
            instructions.push(...this.generateExpression(stmt.value));
        }

        instructions.push(new GaiaInstruction('Return'));

        return instructions;
    }

    private generateIfStatement(stmt: IfStatement): GaiaInstruction[] {
        const instructions: GaiaInstruction[] = [];

        // 生成条件表达式
        instructions.push(...this.generateExpression(stmt.condition));

        // 创建跳转标签
        const elseLabel = `else_${Date.now()}`;
        const endLabel = `end_${Date.now()}`;

        // 条件跳转到else分支
        instructions.push(new GaiaInstruction('JumpIfFalse', [elseLabel]));

        // 生成then分支
        for (const thenStmt of stmt.thenBranch) {
            instructions.push(...this.generateStatement(thenStmt));
        }

        // 跳转到结束
        instructions.push(new GaiaInstruction('Jump', [endLabel]));

        // 生成else分支
        instructions.push(new GaiaInstruction('Label', [elseLabel]));

        if (stmt.elseBranch) {
            for (const elseStmt of stmt.elseBranch) {
                instructions.push(...this.generateStatement(elseStmt));
            }
        }

        // 结束标签
        instructions.push(new GaiaInstruction('Label', [endLabel]));

        return instructions;
    }

    private generateWhileStatement(stmt: WhileStatement): GaiaInstruction[] {
        const instructions: GaiaInstruction[] = [];

        // 创建循环标签
        const startLabel = `loop_start_${Date.now()}`;
        const endLabel = `loop_end_${Date.now()}`;

        // 循环开始标签
        instructions.push(new GaiaInstruction('Label', [startLabel]));

        // 生成条件表达式
        instructions.push(...this.generateExpression(stmt.condition));

        // 条件跳转到结束
        instructions.push(new GaiaInstruction('JumpIfFalse', [endLabel]));

        // 生成循环体
        for (const bodyStmt of stmt.body) {
            instructions.push(...this.generateStatement(bodyStmt));
        }

        // 跳回循环开始
        instructions.push(new GaiaInstruction('Jump', [startLabel]));

        // 循环结束标签
        instructions.push(new GaiaInstruction('Label', [endLabel]));

        return instructions;
    }

    private generateExpressionStatement(stmt: ExpressionStatement): GaiaInstruction[] {
        const instructions = this.generateExpression(stmt.expression);

        // 如果表达式有结果但未被使用，弹出它
        if (stmt.expression.kind !== 'CallExpression' &&
            stmt.expression.kind !== 'Assignment') {
            instructions.push(new GaiaInstruction('Pop'));
        }

        return instructions;
    }

    private generateExpression(expr: Expression): GaiaInstruction[] {
        switch (expr.kind) {
            case 'Literal':
                return this.generateLiteral(expr);
            case 'Identifier':
                return this.generateIdentifier(expr);
            case 'BinaryExpression':
                return this.generateBinaryExpression(expr);
            case 'UnaryExpression':
                return this.generateUnaryExpression(expr);
            case 'CallExpression':
                return this.generateCallExpression(expr);
            default:
                throw new Error(`Unsupported expression kind: ${expr.kind}`);
        }
    }

    private generateLiteral(expr: Literal): GaiaInstruction[] {
        return [new GaiaInstruction('Push', [literalToGaiaConstant(expr)])];
    }

    private generateIdentifier(expr: Identifier): GaiaInstruction[] {
        const index = this.localVariables.get(expr.name);

        if (index === undefined) {
            throw new Error(`Undefined variable: ${expr.name}`);
        }

        return [new GaiaInstruction('LoadLocal', [index])];
    }

    private generateBinaryExpression(expr: BinaryExpression): GaiaInstruction[] {
        const instructions: GaiaInstruction[] = [];

        // 生成左右操作数
        instructions.push(...this.generateExpression(expr.left));
        instructions.push(...this.generateExpression(expr.right));

        // 根据运算符生成对应的指令
        switch (expr.operator) {
            case 'Add':
                instructions.push(new GaiaInstruction('Add'));
                break;
            case 'Subtract':
                instructions.push(new GaiaInstruction('Subtract'));
                break;
            case 'Multiply':
                instructions.push(new GaiaInstruction('Multiply'));
                break;
            case 'Divide':
                instructions.push(new GaiaInstruction('Divide'));
                break;
            case 'Equal':
                instructions.push(new GaiaInstruction('CompareEqual'));
                break;
            case 'NotEqual':
                instructions.push(new GaiaInstruction('CompareEqual'));
                instructions.push(new GaiaInstruction('Not'));
                break;
            case 'LessThan':
                instructions.push(new GaiaInstruction('CompareLess'));
                break;
            case 'LessThanOrEqual':
                instructions.push(new GaiaInstruction('CompareLessOrEqual'));
                break;
            case 'GreaterThan':
                instructions.push(new GaiaInstruction('CompareGreater'));
                break;
            case 'GreaterThanOrEqual':
                instructions.push(new GaiaInstruction('CompareGreaterOrEqual'));
                break;
            default:
                throw new Error(`Unsupported binary operator: ${expr.operator}`);
        }

        return instructions;
    }

    private generateUnaryExpression(expr: UnaryExpression): GaiaInstruction[] {
        const instructions: GaiaInstruction[] = [];

        // 生成操作数
        instructions.push(...this.generateExpression(expr.operand));

        // 根据运算符生成对应的指令
        switch (expr.operator) {
            case 'Negate':
                instructions.push(new GaiaInstruction('Negate'));
                break;
            case 'Not':
                instructions.push(new GaiaInstruction('Not'));
                break;
            default:
                throw new Error(`Unsupported unary operator: ${expr.operator}`);
        }

        return instructions;
    }

    private generateCallExpression(expr: CallExpression): GaiaInstruction[] {
        const instructions: GaiaInstruction[] = [];

        // 生成参数
        for (const arg of expr.arguments) {
            instructions.push(...this.generateExpression(arg));
        }

        // 生成函数调用
        instructions.push(new GaiaInstruction('Call', [expr.callee.name]));

        return instructions;
    }
}

File: examples/mini-rust/src/codegen.rs
//! Mini Rust 代码生成器

use crate::ast::*;
use gaia_assembler::*;
use gaia_types::*;
use std::collections::HashMap;

/// 代码生成器
pub struct CodeGenerator {
    /// 局部变量映射
    locals: HashMap<String, usize>,
    /// 当前局部变量索引
    local_index: usize,
}

impl CodeGenerator {
    pub fn new() -> Self {
        Self { locals: HashMap::new(), local_index: 0 }
    }

    /// 生成 GaiaProgram
    pub fn generate(&mut self, program: &Program) -> Result<GaiaProgram> {
        let mut functions = Vec::new();

        for function in &program.functions {
            functions.push(self.generate_function(function)?);
        }

        Ok(GaiaProgram { name: program.name.clone(), functions, constants: Vec::new() })
    }

    /// 生成函数
    fn generate_function(&mut self, function: &Function) -> Result<GaiaFunction> {
        // 重置局部变量状态
        self.locals.clear();
        self.local_index = 0;

        // 添加参数到局部变量
        for param in &function.parameters {
            self.locals.insert(param.name.clone(), self.local_index);
            self.local_index += 1;
        }

        let mut instructions = Vec::new();

        // 生成函数体
        self.generate_block(&function.body, &mut instructions)?;

        // 如果函数没有显式返回，添加默认返回
        if !instructions.iter().any(|inst| matches!(inst, GaiaInstruction::Return)) {
            instructions.push(GaiaInstruction::Return);
        }

        Ok(GaiaFunction {
            name: function.name.clone(),
            parameters: function.parameters.iter().map(|p| p.param_type.to_gaia_type()).collect(),
            return_type: function.return_type.as_ref().map(|t| t.to_gaia_type()),
            locals: Vec::new(), // 暂时为空，后续可以优化
            instructions,
        })
    }

    /// 生成代码块
    fn generate_block(&mut self, block: &Block, instructions: &mut Vec<GaiaInstruction>) -> Result<()> {
        for statement in &block.statements {
            self.generate_statement(statement, instructions)?;
        }
        Ok(())
    }

    /// 生成语句
    fn generate_statement(&mut self, statement: &Statement, instructions: &mut Vec<GaiaInstruction>) -> Result<()> {
        match statement {
            Statement::Expression(expr) => {
                self.generate_expression(expr, instructions)?;
                // 表达式语句需要弹出结果
                instructions.push(GaiaInstruction::Pop);
            }
            Statement::VariableDeclaration { name, var_type: _, initializer } => {
                if let Some(init_expr) = initializer {
                    // 生成初始值
                    self.generate_expression(init_expr, instructions)?;
                }
                else {
                    // 如果没有初始值，使用默认值 0
                    instructions.push(GaiaInstruction::LoadConstant(GaiaConstant::Integer32(0)));
                }

                // 分配局部变量
                let index = self.local_index;
                self.locals.insert(name.clone(), index);
                self.local_index += 1;

                // 存储到局部变量
                instructions.push(GaiaInstruction::StoreLocal(index as u32));
            }
            Statement::Return(expr) => {
                if let Some(expr) = expr {
                    self.generate_expression(expr, instructions)?;
                }
                instructions.push(GaiaInstruction::Return);
            }
        }
        Ok(())
    }

    /// 生成表达式
    fn generate_expression(&mut self, expression: &Expression, instructions: &mut Vec<GaiaInstruction>) -> Result<()> {
        match expression {
            Expression::Literal(literal) => {
                let constant = literal.to_gaia_constant();
                instructions.push(GaiaInstruction::LoadConstant(constant));
            }
            Expression::Identifier(name) => {
                if let Some(&index) = self.locals.get(name) {
                    instructions.push(GaiaInstruction::LoadLocal(index as u32));
                }
                else {
                    return Err(GaiaError::syntax_error(format!("未定义的变量: {}", name), SourceLocation::default()));
                }
            }
            Expression::BinaryOperation { left, operator, right } => {
                self.generate_expression(left, instructions)?;
                self.generate_expression(right, instructions)?;

                let instruction = match operator {
                    BinaryOperator::Add => GaiaInstruction::Add,
                    BinaryOperator::Subtract => GaiaInstruction::Subtract,
                    BinaryOperator::Multiply => GaiaInstruction::Multiply,
                    BinaryOperator::Divide => GaiaInstruction::Divide,
                    BinaryOperator::Equal => GaiaInstruction::CompareEqual,
                    BinaryOperator::NotEqual => {
                        // 先比较相等，然后取反
                        instructions.push(GaiaInstruction::CompareEqual);
                        instructions.push(GaiaInstruction::LoadConstant(GaiaConstant::Integer32(0)));
                        instructions.push(GaiaInstruction::CompareEqual);
                        return Ok(());
                    }
                    BinaryOperator::Less => GaiaInstruction::CompareLessThan,
                    BinaryOperator::LessEqual => {
                        // a <= b 等价于 !(a > b)
                        instructions.push(GaiaInstruction::CompareGreaterThan);
                        instructions.push(GaiaInstruction::LoadConstant(GaiaConstant::Integer32(0)));
                        instructions.push(GaiaInstruction::CompareEqual);
                        return Ok(());
                    }
                    BinaryOperator::Greater => GaiaInstruction::CompareGreaterThan,
                    BinaryOperator::GreaterEqual => {
                        // a >= b 等价于 !(a < b)
                        instructions.push(GaiaInstruction::CompareLessThan);
                        instructions.push(GaiaInstruction::LoadConstant(GaiaConstant::Integer32(0)));
                        instructions.push(GaiaInstruction::CompareEqual);
                        return Ok(());
                    }
                };

                instructions.push(instruction);
            }
            Expression::UnaryOperation { operator, operand } => {
                match operator {
                    UnaryOperator::Negate => {
                        // 数值取负：0 - operand
                        instructions.push(GaiaInstruction::LoadConstant(GaiaConstant::Integer32(0)));
                        self.generate_expression(operand, instructions)?;
                        instructions.push(GaiaInstruction::Subtract);
                    }
                    UnaryOperator::Not => {
                        // 逻辑非：operand == 0
                        self.generate_expression(operand, instructions)?;
                        instructions.push(GaiaInstruction::LoadConstant(GaiaConstant::Integer32(0)));
                        instructions.push(GaiaInstruction::CompareEqual);
                    }
                }
            }
            Expression::FunctionCall { name, arguments } => {
                // 生成参数
                for arg in arguments {
                    self.generate_expression(arg, instructions)?;
                }

                // 处理内置函数
                if name == "console" {
                    return Err(GaiaError::syntax_error("console 不是函数，应该使用 console.log()", SourceLocation::default()));
                }

                // 调用函数
                instructions.push(GaiaInstruction::Call(name.clone()));
            }
            Expression::MethodCall { object, method, arguments } => {
                // 处理 console.log
                if let Expression::Identifier(obj_name) = object.as_ref() {
                    if obj_name == "console" && method == "log" {
                        // 生成参数
                        for arg in arguments {
                            self.generate_expression(arg, instructions)?;
                        }

                        // 调用内置的打印函数
                        instructions.push(GaiaInstruction::Call("__builtin_print".to_string()));
                        return Ok(());
                    }
                }

                // 其他方法调用暂不支持
                return Err(GaiaError::syntax_error(
                    format!("不支持的方法调用: {}.{}", self.expression_to_string(object), method),
                    SourceLocation::default(),
                ));
            }
            Expression::MacroCall { name, arguments } => {
                // 处理 println! 宏
                if name == "println" {
                    // 生成参数
                    for arg in arguments {
                        self.generate_expression(arg, instructions)?;
                    }

                    // 调用内置的打印函数
                    instructions.push(GaiaInstruction::Call("__builtin_println".to_string()));
                }
                else {
                    return Err(GaiaError::syntax_error(format!("不支持的宏: {}!", name), SourceLocation::default()));
                }
            }
        }
        Ok(())
    }

    /// 将表达式转换为字符串（用于错误消息）
    fn expression_to_string(&self, expr: &Expression) -> String {
        match expr {
            Expression::Identifier(name) => name.clone(),
            Expression::Literal(Literal::String(s)) => format!("\"{}\"", s),
            Expression::Literal(Literal::Integer(i)) => i.to_string(),
            Expression::Literal(Literal::Float(f)) => f.to_string(),
            Expression::Literal(Literal::Boolean(b)) => b.to_string(),
            _ => "<expression>".to_string(),
        }
    }
}

/// Mini Rust 解析器
pub struct MiniRustParser;

impl MiniRustParser {
    pub fn parse(source: &str) -> Result<GaiaProgram> {
        use crate::{lexer::RustLexer, parser::Parser};

        let mut lexer = RustLexer::new(source);
        let diagnostics = lexer.tokenize();
        let token_stream = diagnostics.result?;

        // 调试：打印所有 tokens
        println!("Generated tokens:");
        let tokens_vec = token_stream.tokens.get_ref();
        for (i, token) in tokens_vec.iter().enumerate() {
            println!("  {}: {:?}", i, token);
        }

        let mut parser = Parser::new(token_stream);
        let ast = parser.parse_program()?;

        let mut codegen = CodeGenerator::new();
        codegen.generate(&ast)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_simple_function() {
        let source = r#"
            fn main() {
                console.log("Hello Gaia!");
            }
        "#;

        let program = MiniRustParser::parse(source).unwrap();
        assert_eq!(program.functions.len(), 1);
        assert_eq!(program.functions[0].name, "main");
    }

    #[test]
    fn test_arithmetic() {
        let source = r#"
            fn add() {
                let result = 1 + 2;
                console.log(result);
            }
        "#;

        let program = MiniRustParser::parse(source).unwrap();
        assert_eq!(program.functions.len(), 1);

        // 检查指令序列
        let instructions = &program.functions[0].instructions;
        assert!(instructions.contains(&GaiaInstruction::Ldc(GaiaConstant::I32(1))));
        assert!(instructions.contains(&GaiaInstruction::Ldc(GaiaConstant::I32(2))));
        assert!(instructions.contains(&GaiaInstruction::Add));
    }
}
